### PR TITLE
[M380] Refactor offlineservice and include profile id in table names

### PR DESCRIPTION
**Pull request by gridcell**

### CODE REVIEW

**Comment by gridcell** [comment]
[handle multiple users on same device](https://trello.com/c/eJ2xI8iX/380-handle-multiple-users-on-same-device)

File: src/app/_mermaid_common/libs/services/offline-tables.svc.js [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+      if (angular.isArray(baseNames) === false) {
+        baseNames = [baseNames];
+        isMulti = false;
+      }
+
+      return authService.getProfileId().then(function(profileId) {
+        const results = baseNames.map(function(baseName) {
+          return `${
+            APP_CONFIG.localDbName
+          }${OfflineTableUtils.TABLE_NAME_DELIMITER}${baseName}${OfflineTableUtils.TABLE_NAME_DELIMITER}${profileId}${OfflineTableUtils.TABLE_NAME_DELIMITER}${projectId}`;
```
Comment (gridcell) [inline src/app/_mermaid_common/libs/services/offline-tables.svc.js]: profileid and project id should be swapped

File: src/app/_mermaid_common/libs/services/offline-table-utils.svc.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        }
+      });
+    };
+
+    const paginatedRefresh = function(table, options) {
+      var limit = options.limit || 100;
+      return _refresh(table, limit);
+    };
+
+    const projectIdFromTableName = function(tableName) {
```
Comment (gridcell) [inline src/app/_mermaid_common/libs/services/offline-table-utils.svc.js]: check if this is still needed

File: src/app/_mermaid_common/libs/services/offline-tables.svc.js [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    const COLLECT_RECORDS_NAME = 'collectrecords';
+
+    const PROJECT_TABLE_NAMES = [
+      PROJECT_SITES_NAME,
+      PROJECT_MANAGEMENTS_NAME,
+      PROJECT_PROFILES_NAME,
+      COLLECT_RECORDS_NAME
+    ];
+
+    const getProjectsTableName = function(excludeProfileId) {
```
Comment (gridcell) [inline src/app/_mermaid_common/libs/services/offline-tables.svc.js]: check if `excludeProfileId` is still needed

File: src/app/app.run.js [hunk 4]
```diff
@@ -33,12 +34,15 @@
...
     layoutUtils,
     system,
+    offlineservice,
     // $urlRouter,
     ConnectivityFactory
   ) {
     const conn = new ConnectivityFactory($rootScope);
     let handleAuth;
 
+    offlineservice.syncAndDeleteV1Tables();
```
Comment (gridcell) [inline src/app/app.run.js]: add message if unable to sync

File: src/app/auth/auth.service.js [hunk 5]
```diff
@@ -131,6 +132,17 @@ angular.module('app.auth').service('authService', [
       return $q.resolve(null);
     }
 
+    function getProfileId() {
+      profileIdPromise = getCurrentUser()
```
Comment (gridcell) [inline src/app/auth/auth.service.js]: profile promise should return promise if not null

File: src/app/project/module.js [hunk 6]
```diff
@@ -67,29 +67,35 @@ angular
       });
     };
 
-    const _getMyProjectProfile = function($stateParams, ProjectService) {
+    const _getMyProjectProfile = function($q, $stateParams, ProjectService) {
       return ProjectService.getMyProjectProfile($stateParams.project_id);
+      // .then(function(projectProfile) {
```
Comment (gridcell) [inline src/app/project/module.js]: Check why this is commented out, remove commented out code if it's not needed

File: src/app/_mermaid_common/libs/services/offline-common-tables.svc.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    'use strict';
+
+    const CHOICES_NAME = 'choices';
+    // const FISH_ATTRIBUTES_NAME = 'fishattributes';
+    const FISH_SIZES_NAME = 'fishsizes';
+    const FISH_FAMILIES_NAME = 'fishfamilies';
+    const FISH_GENERA_NAME = 'fishgenera';
+    const FISH_SPECIES_NAME = 'fishspecies';
+    const BENTHIC_ATTRIBUTES_NAME = 'benthicattributes';
+    // const PROJECT_TAGS_NAME = 'projecttags';
```
[the remark this answers is not in the document]
Comment (gridcell) [reply src/app/_mermaid_common/libs/services/offline-common-tables.svc.js]: project tags are only accessed online, not needed

File: src/app/_mermaid_common/libs/services/offline-tables.svc.js (hunk 1 above)
Comment (gridcell) [reply src/app/_mermaid_common/libs/services/offline-tables.svc.js]: this is correct the way it is

File: src/app/project/module.js (hunk 6 above)
Comment (gridcell) [reply src/app/project/module.js]: Didn't see why we don't want to do this, I uncommented it

File: src/app/_mermaid_common/libs/factories/offline-table.factory.js [hunk 8]
```diff
@@ -61,13 +61,77 @@
...
+    'use strict';
+    const listeners = {};
+
+    const addListener = function(uniqueId, callback, context, keyName) {
+      let _cb = callback;
+      let _context = context;
+      let _keyName = keyName;
+      let isNew = true;
+
+      if (keyName != null) {
```
Comment (saanobhaai) [inline src/app/_mermaid_common/libs/factories/offline-table.factory.js]: What happens if `keyName == null`?

**Pull request merged**